sklearn/neighbors/base.py

```diff
@@ -164,41 +163,31 @@ def _kneighbors_from_graph(graph, n_neighbors):
     """
     n_samples = graph.shape[0]
 
-    # number of neighbors for each samples
+    # number of neighbors by samples
     row_nnz = np.diff(graph.indptr)
-    if row_nnz.min() < n_neighbors:
+    row_nnz_min = row_nnz.min()
+    if n_neighbors is not None and row_nnz_min < n_neighbors:
         raise ValueError(
             '%d neighbors per samples are required, but some samples have only'
             ' %d neighbors in precomputed graph matrix. Decrease number of '
             'neighbors used or recompute the graph with more neighbors.'
-            % (n_neighbors, row_nnz.min()))
+            % (n_neighbors, row_nnz_min))
 
     # if there is the same number of neighbors for each sample
-    if row_nnz.max() == row_nnz.min():
-        neigh_dist = graph.data.reshape(n_samples, -1)
-        neigh_ind = graph.indices.reshape(n_samples, -1)
-
-        # sort neigh_dist and neigh_ind to have increasing distances
-        n_samples, n_neighbors_2 = neigh_dist.shape
-        perm = np.argsort(neigh_dist)
-        perm += np.arange(n_samples)[:, None] * n_neighbors_2
-        perm = perm.ravel()
-        neigh_dist = neigh_dist.ravel()[perm].reshape(n_samples, n_neighbors_2)
-        neigh_ind = neigh_ind.ravel()[perm].reshape(n_samples, n_neighbors_2)
-        return neigh_dist[:, :n_neighbors], neigh_ind[:, :n_neighbors]
+    if row_nnz.max() == row_nnz_min:
 
-    else:
-        neigh_dist = np.empty((n_samples, n_neighbors))
-        neigh_ind = np.empty((n_samples, n_neighbors))
-        for i in range(n_samples):
-            idx = slice(graph.indptr[i], graph.indptr[i + 1])
-            this_dist = graph.data[idx]
-            this_ind = graph.indices[idx]
-            # sort neigh_dist and neigh_ind to have increasing distances
-            perm = np.argsort(this_dist)
-            neigh_dist[i] = this_dist[perm][:n_neighbors]
-            neigh_ind[i] = this_ind[perm][:n_neighbors]
-        return neigh_dist, neigh_ind
+        def extract(a):
+            return a.reshape(n_samples, -1)[:, :n_neighbors]
+
+    else:  # radius neighbors case
```

Comment:
I don't think you should call this "radius neighbors case". "Each sample has different number of provided neighbors" is much more explicit and clear.